src/browser/browser.zig

```diff
@@ -380,7 +364,37 @@ pub const Page = struct {
 
         var response = try request.sendSync(.{});
         const header = response.header;
-        try self.session.cookie_jar.populateFromResponse(self.uri, &header);
+        try self.session.cookie_jar.populateFromResponse(request.uri, &header);
+
+        // update uri after eventual redirection
+        var buf: std.ArrayListUnmanaged(u8) = .{};
+        defer buf.deinit(arena);
+
+        buf.clearRetainingCapacity();
+        try request.uri.writeToStream(.{
+            .scheme = true,
+            .authentication = true,
+            .authority = true,
+            .path = true,
+            .query = true,
+            .fragment = true,
+        }, buf.writer(arena));
+        self.rawuri = try buf.toOwnedSlice(arena);
```

Comment:
I see that you re-use the buffer later, which I guess is why you copy the contents. But, `toOwnedSlice` frees the underlying memory, so the underlying slice won't get re-used.  I think you can use `buf.items` to avoid the allocation + copy.